packages/frontend/core/src/desktop/pages/open-app/index.tsx

```diff
@@ -1,18 +1,48 @@
 import { useNavigateHelper } from '@affine/core/components/hooks/use-navigate-helper';
 import { GraphQLService } from '@affine/core/modules/cloud';
 import { OpenInAppPage } from '@affine/core/modules/open-in-app/views/open-in-app-page';
-import { appSchemes, channelToScheme } from '@affine/core/utils/channel';
+import {
+  appSchemaUrl,
+  appSchemes,
+  channelToScheme,
+} from '@affine/core/utils/channel';
+import { extractUrlSearchParams } from '@affine/core/utils/url';
 import type { GetCurrentUserQuery } from '@affine/graphql';
 import { getCurrentUserQuery } from '@affine/graphql';
 import { useService } from '@toeverything/infra';
 import { useCallback, useEffect, useState } from 'react';
-import { useParams, useSearchParams } from 'react-router-dom';
+import {
+  type LoaderFunction,
+  redirect,
+  useLoaderData,
+  useParams,
+} from 'react-router-dom';
+import { z } from 'zod';
 
 import { AppContainer } from '../../components/app-container';
 
+const LoaderDataSchema = z.object({
+  action: z.enum(['url', 'signin-redirect']),
```

Comment:
action doesn't seem to need judgment, filtering is already done inside the `Component`, illegal actions cannot be rendered

Reply:
do not use loader anymore. we are gonna switch to react-router v7 library mode